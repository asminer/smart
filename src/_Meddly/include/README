
# $Id$

(Incomplete)

Explanation of files:

meddly.h          simple interface
meddly_expert.h   expert interface

Usage:
meddly.h must be included before meddly_expert.h.

