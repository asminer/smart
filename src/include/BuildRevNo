#!/bin/bash

#
# $Id$
#
# Builds the revision number.
#

if [ $# -ne 1 ]; then
  echo Usage: $0 "<repospath>"
  exit
fi

path=$1
revfile=$path/revision.h

rev=`svn info $path | awk '/Revision/{print $2}'`
if [ -z "$rev" ]; then
  if [ -e $revfile ]; then
    printf "Leaving old revision number\n"
  else
    printf "Using bogus revision number\n";
    printf "const int REVISION_NUMBER = 0 ;\n" > $revfile
  fi
else
  if [ -e $revfile ]; then
    revlast=`awk '{print $5}' $revfile`
    if [ "$rev" -eq "$revlast" ]; then
      printf "No revision number change since last build\n";
    else
      printf "Updating revision number to %d\n" $rev
      printf "const int REVISION_NUMBER = $rev ;\n" > $revfile
    fi
  else
    printf "Updating revision number to %d\n" $rev
    printf "const int REVISION_NUMBER = $rev ;\n" > $revfile
  fi
fi

