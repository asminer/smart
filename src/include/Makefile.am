## $Id$

## Makefile.am -- Process this file with automake to produce Makefile.in

_revision:
	$(srcdir)/BuildRevNo $(srcdir) 

revision.h:
	$(srcdir)/BuildRevNo $(srcdir)

BUILT_SOURCES = _revision revision.h
CLEANFILES = revision.h

EXTRA_DIST = BuildRevNo revision.h
