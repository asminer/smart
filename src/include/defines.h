
// $Id$

/** @name defines.h
    @type File
    @args \ 

    The base of all modules.  So if you change this, everything gets
    to recompile.

    This file is for good global defines, such as DCASSERT.

    Since this file is only intended for global definitions, there is no
    associated defines.c or defines.cc file.

 */

//@{

#ifndef DEFINES_H
#define DEFINES_H

#include <assert.h>

// Handy Macros

/// Standard MAX "macro".
template <class T> inline T MAX(T X,T Y) { return ((X>Y)?X:Y); }
/// Standard MIN "macro".
template <class T> inline T MIN(T X,T Y) { return ((X<Y)?X:Y); }
/// Standard ABS "macro".
template <class T> inline T ABS(T X) { return ((X<0)?(-X):(X)); }
/// SWAP "macro".
template <class T> inline void SWAP(T &x, T &y) { T tmp=x; x=y; y=tmp; }
/// POSITIVE "macro".
template <class T> inline bool POSITIVE(T X) { return (X>0) ? 1 : 0; }
/// SIGN "macro".
template <class T> inline int SIGN(T X) { return (X<0) ? -1 : POSITIVE(X); }

/*
    There are now two modes of code generation:
       "DEVELOPMENT_CODE" and "RELEASE_CODE".
    
    If "DEVELOPMENT_CODE" is defined (usually done in the makefile) then 
    debugging macros and assertions will be turned on.  Otherwise we assume
    that we have "RELEASE_CODE" and they are turned off.
    
    Macros useful for debugging "development code" that are turned off 
    for release code (for speed):

    DCASSERT()
    CHECK_RANGE(low, x, high+1)
*/

#ifdef DEVELOPMENT_CODE
  #define MEM_TRACE_ON  
  #define RANGE_CHECK_ON
  #define DCASSERTS_ON
#endif


// Use this for assertions that will fail only when your
// code is wrong.  Handy for debugging.
#ifdef DCASSERTS_ON 
#define DCASSERT(X) assert(X)
#else
#define DCASSERT(X)
#endif

// Also useful for debugging.
#ifdef RANGE_CHECK_ON
  inline void CheckRange(int min, int value, int max)
  {
    assert(value<max);
    assert(value>=min);
  }
  #define CHECK_RANGE(MIN, VALUE, MAX)  CheckRange(MIN, VALUE, MAX)
#else
  #define CHECK_RANGE(MIN, VALUE, MAX)
#endif

// More debugging tricks
#ifdef DEVELOPMENT_CODE
  #define CHECK_RETURN(X, V)  DCASSERT(V == X)
#else
  #define CHECK_RETURN(X, V)  X
#endif

// Safe typecasting for development code;  fast casting otherwise

#ifdef DEVELOPMENT_CODE
#define smart_cast  dynamic_cast
#else
#define smart_cast  static_cast
#endif

// Flip this on for malloc / free errors

// #define DEBUG_MEM

#endif

//@}

