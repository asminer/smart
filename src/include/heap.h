
// $Id$

#ifndef HEAP_H
#define HEAP_H

const int MAX_HEAP_ADD = 1024;


// ==================================================================
// ||                                                              ||
// ||                     HeapOfPointers class                     ||
// ||                                                              ||
// ==================================================================


/**  Heap template class, as in HeapSort.
     The heap stores pointers to objects.
     To use, there must be a function
        int Compare(DATA *a, DATA *b) that works like strcmp.
     Note: the heap keeps the LARGEST element at the root,
     so that the sorted array is in INCREASING order.

     Heap method "show" requires a similar method for the stored objects.
*/

template <class DATA>
class HeapOfPointers {
  DATA **data;  
  long last;
  long size;
  bool sorted;
protected:
  bool Resize(long newsize) {
    if (newsize == size) return true;
    DATA ** foo = (DATA**) realloc(data, newsize*sizeof(void*));
    if (newsize && (0==foo)) return false;
    data = foo;
    size = newsize;
    return true;
  }
public:
  /// Constructor.  Initially empty and zero size.
  HeapOfPointers() {
    data = 0;
    size = 0;
    last = 0;
    sorted = false;
  }

  /// Constructor.  Specify the initial array size.
  HeapOfPointers(long inits) {
    data = 0;
    size = 0;
    Resize(inits);
    last = 0;
    sorted = false;
  }
  /// Constructor, give initial unsorted array.
  HeapOfPointers(DATA **a, long sz) {
    data = a;
    size = sz;
    sorted = false;
    // make this array a proper heap
    last = 0;
    while (last<size) {
      UpHeap(last);
      last++; 
    } 
  }
  ~HeapOfPointers() {
    Resize(0);
  }
  void Clear(long inits) {
    last = 0;
    sorted = false;
    if (inits>size) Resize(inits);
  }
  inline long Length() { return last; }
  /// Returns true on success, false on out of memory error.
  bool Insert(DATA* a) {
    DCASSERT(!sorted);
    if (last>=size) {
      long newsize = MIN(2*size, size+MAX_HEAP_ADD);
      if (newsize < 1)  newsize = 16;
      if (!Resize(newsize))
      return false;
    }
    data[last] = a;
    UpHeap(last);
    last++;
    return true;
  }
  void Remove(DATA* &a) {
    DCASSERT(!sorted);
    Swap(0, last-1);
    DownHeap(last-1);
    last--;
    a = data[last]; 
  }
  void Sort();
  DATA** MakeArray() {
    DCASSERT(sorted);
    Resize(last);
    DATA** ret = data;
    size = last = 0;
    data = 0;
    sorted = false;
    return ret;
  }
  // for debugging
  void show(OutputStream &s) const;
protected:
  inline void Swap(long i, long j) {
    if (i!=j) {
      DATA *t = data[i];
      data[i] = data[j];
      data[j] = t;
    }
  }
  void UpHeap(long n);
  void DownHeap(long n);
};

template <class DATA>
void HeapOfPointers<DATA>::Sort()
{
  DCASSERT(!sorted);
  long n;
  for (n=last-1; n>0; n--) {
    Swap(0, n);
    DownHeap(n);
  }
  sorted = true;
}

template <class DATA>
void HeapOfPointers<DATA>::UpHeap(long n)
{
  while (n) {
    long parent = (n-1)/2;
    int cmp = Compare(data[n], data[parent]);
    if (cmp<=0) return; // child <= parent, done
    Swap(n, parent);
    n = parent;
  }
}

template <class DATA>
void HeapOfPointers<DATA>::DownHeap(long n)
{
  long x=0;
  while (1) {
    long a = 2*x+1;  // left child
    long b = a+1;  // right child
    if (a>=n) return;  // no children
    if (b>=n) {
      // this node has one left child (a leaf), and no right child.
      // check the child and swap if necessary, then we're done
      int cmp = Compare(data[a], data[x]);
      if (cmp>0) Swap(a,x); // a > x
      return;
    }
    bool xbeatsa = (Compare(data[x], data[a]) >= 0);
    bool xbeatsb = (Compare(data[x], data[b]) >= 0);
    if (xbeatsa && xbeatsb) return;  // already a heap
    if (xbeatsa) { // b is the new parent
      Swap(b,x);
      x = b; 
      continue;
    }
    if (xbeatsb) { // a is the new parent
      Swap(a,x);
      x = a;
      continue;
    }
    // Still here?  we must have a>x and b>x; 
    // New parent is max of a and b.
    if (Compare(data[a], data[b]) >= 0) {
      Swap(a,x); // a is new parent
      x = a;
    } else {
      Swap(b,x); // b is new parent
      x = b;
    }
  } // while
}

template <class DATA>
void HeapOfPointers<DATA>::show(OutputStream &s) const
{
  s << "Heap, by node:\n";
  long i;
  for (i=0; i<last; i++) {
    s << "Node " << i << ":\n";
    s << "\tobject: " << data[i] << "\n";
    long lc = 2*i+1;
    long rc = lc+1;
    if (lc<last)
      s << "\t  left: " << data[lc] << "\n";
    if (rc<last)
      s << "\t right: " << data[rc] << "\n";
    s.flush(); 
  }
  s << "End of heap\n";
  s.flush();
}



// ==================================================================
// ||                                                              ||
// ||              HeapSort functions, array of items              ||
// ||                                                              ||
// ==================================================================

/** Standard "UpHeap" procedure for an array of items.
    The array itself is modified.
    PRE: items 0 through n-1 already form a heap.
    POST: items 0 through n form a heap.
*/
template <class DATA>
void UpHeap(DATA* D, long n)
{
  while (n) {
    long parent = (n-1)/2;
    if (D[n] <= D[parent])
      return; // child <= parent, done
    SWAP(D[n], D[parent]);
    n = parent;
  }
}

/** Standard "DownHeap" procedure for an array of items.
    The array itself is modified.
    PRE: items 1 through n-1 already form a heap.
    POST: items 0 through n-1 form a heap.
*/
template <class DATA>
void DownHeap(DATA* D, long n)
{
  long x=0;
  while (1) {
    long a = 2*x+1;  // left child
    long b = a+1;  // right child
    if (a>=n) return;  // no children
    if (b>=n) {
      // this node has one left child (a leaf), and no right child.
      // check the child and swap if necessary, then we're done
      if (D[a] > D[x])
          SWAP(D[a], D[x]);  // a > x
      return;
    }
    bool xbeatsa = D[x] >= D[a];
    bool xbeatsb = D[x] >= D[b];
    if (xbeatsa && xbeatsb) return;  // already a heap
    if (xbeatsa) { // b is the new parent
      SWAP(D[x], D[b]);
      x = b; 
      continue;
    }
    if (xbeatsb) { // a is the new parent
      SWAP(D[x], D[a]);
      x = a;
      continue;
    }
    // Still here?  we must have a>x and b>x; 
    // New parent is max of a and b.
    if (D[a] >= D[b]) {
      SWAP(D[x], D[a]); // a is new parent
      x = a;
    } else {
      SWAP(D[x], D[b]); // b is new parent
      x = b;
    }
  } // while
}

/** Heapsort for an array of items.
    The array itself is modified.
    On return, data[i] gives the ith smallest item.
*/
template <class DATA>
void HeapSort(DATA* data, long N)
{
  long i;
  for (i=1; i<N; i++) {
    UpHeap(data, i);
  }
  for (i--; i>0; i--) {
    SWAP(data[0], data[i]);
    DownHeap(data, i);
  }
}

// ==================================================================
// ||                                                              ||
// ||     HeapSort functions,  array of items with permutation     ||
// ||                                                              ||
// ==================================================================

/** Standard "UpHeap" procedure for an array of items.
    The array itself is not modified, we use a permutation array instead.
    PRE: items 0 through n-1 already form a heap.
    POST: items 0 through n form a heap.
*/
template <class DATA>
void UpHeap(const DATA* D, long* perm, long n)
{
  while (n) {
    long parent = (n-1)/2;
    if (D[perm[n]] <= D[perm[parent]])
      return; // child <= parent, done
    SWAP(perm[n], perm[parent]);
    n = parent;
  }
}

/** Standard "DownHeap" procedure for an array of items.
    The array itself is not modified, we use a permutation array instead.
    PRE: items 1 through n-1 already form a heap.
    POST: items 0 through n-1 form a heap.
*/
template <class DATA>
void DownHeap(const DATA* D, long* perm, long n)
{
  long x=0;
  while (1) {
    long a = 2*x+1;  // left child
    long b = a+1;  // right child
    if (a>=n) return;  // no children
    if (b>=n) {
      // this node has one left child (a leaf), and no right child.
      // check the child and swap if necessary, then we're done
      if (D[perm[a]] > D[perm[x]])
          SWAP(perm[a], perm[x]);  // a > x
      return;
    }
    bool xbeatsa = D[perm[x]] >= D[perm[a]];
    bool xbeatsb = D[perm[x]] >= D[perm[b]];
    if (xbeatsa && xbeatsb) return;  // already a heap
    if (xbeatsa) { // b is the new parent
      SWAP(perm[x], perm[b]);
      x = b; 
      continue;
    }
    if (xbeatsb) { // a is the new parent
      SWAP(perm[x], perm[a]);
      x = a;
      continue;
    }
    // Still here?  we must have a>x and b>x; 
    // New parent is max of a and b.
    if (D[perm[a]] >= D[perm[b]]) {
      SWAP(perm[x], perm[a]); // a is new parent
      x = a;
    } else {
      SWAP(perm[x], perm[b]); // b is new parent
      x = b;
    }
  } // while
}

/** Heapsort for an array of items.
    The array itself is not modified, we use a permutation array instead.
    On return, perm[i] gives the index of the ith smallest item in D.
*/
template <class DATA>
void HeapSort(const DATA* data, long* perm, long N)
{
  long i;
  for (i=0; i<N; i++) perm[i] = i;
  for (i=1; i<N; i++) {
    UpHeap(data, perm, i);
  }
  for (i--; i>0; i--) {
    SWAP(perm[0], perm[i]);
    DownHeap(data, perm, i);
  }
}

// ==================================================================
// ||                                                              ||
// ||            HeapSort functions, general collection            ||
// ||                                                              ||
// ==================================================================

/** Standard "UpHeap" procedure for an arbitrary collection.
    PRE: items 0 through n-1 already form a heap.
    POST: items 0 through n form a heap.
    The collection class must provide functions:
      Compare(long i, long j)
      Swap(long i, long j)
    where \a i and \a j are "item numbers".
*/
template <class COLL>
void UpHeapAbstract(COLL* data, long n)
{
  while (n) {
    long parent = (n-1)/2;
    int cmp = data->Compare(n, parent);
    if (cmp<=0) return; // child <= parent, done
    data->Swap(n, parent);
    n = parent;
  }
}

/** Standard "DownHeap" procedure for an arbitrary collection.
    PRE: items 1 through n-1 already form a heap.
    POST: items 0 through n-1 form a heap.
    The collection class must provide functions:
      Compare(long i, long j)
      Swap(long i, long j)
    where \a i and \a j are "item numbers".
*/
template <class COLL>
void DownHeapAbstract(COLL* data, long n)
{
  long x=0;
  while (1) {
    long a = 2*x+1;  // left child
    long b = a+1;  // right child
    if (a>=n) return;  // no children
    if (b>=n) {
      // this node has one left child (a leaf), and no right child.
      // check the child and swap if necessary, then we're done
      int cmp = data->Compare(a, x);
      if (cmp>0) data->Swap(a,x); // a > x
      return;
    }
    bool xbeatsa = (data->Compare(x, a) >= 0);
    bool xbeatsb = (data->Compare(x, b) >= 0);
    if (xbeatsa && xbeatsb) return;  // already a heap
    if (xbeatsa) { // b is the new parent
      data->Swap(b,x);
      x = b; 
      continue;
    }
    if (xbeatsb) { // a is the new parent
      data->Swap(a,x);
      x = a;
      continue;
    }
    // Still here?  we must have a>x and b>x; 
    // New parent is max of a and b.
    if (data->Compare(a, b) >= 0) {
      data->Swap(a,x); // a is new parent
      x = a;
    } else {
      data->Swap(b,x); // b is new parent
      x = b;
    }
  } // while
}

/** Heapsort for an arbitrary collection.
    The collection class must provide functions:
      Compare(long i, long j)
      Swap(long i, long j)
    where \a i and \a j are "item numbers".
*/
template <class COLL>
void HeapSortAbstract(COLL* data, long N)
{
  long i;
  for (i=1; i<N; i++) {
    UpHeapAbstract(data, i);
  }
  for (i--; i>0; i--) {
    data->Swap(0, i);
    DownHeapAbstract(data, i);
  }
}

#endif

