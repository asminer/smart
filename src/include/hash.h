
// $Id$

#ifndef HASH_H
#define HASH_H

#include "defines.h"
#include <stdlib.h> // for realloc

class voidhash {
protected:
  static const long SIZES[];
  void** table;
  int size_index;
  long elements;
  // stats
  long maxchain;
  int max_size_index;
  long maxelems;
  long num_resizes;
public:
  voidhash() {
    size_index = 0;
    max_size_index = 0;
    elements = 0;
    maxelems = 0;
    maxchain = 0;
    num_resizes = 0;
    table = (void**) malloc(getSize() * sizeof(void*));
    for (long i=getSize()-1; i>=0; i--) table[i] = 0;
  }
  ~voidhash() {
    free(table);
  }
  inline long getSize() const { return SIZES[size_index]; }
  inline long getMaxSize() const { return SIZES[max_size_index]; }
  inline long getNumElements() const { return elements; }
  inline long getMaxElements() const { return maxelems; }
  inline long getMaxChain() const { return maxchain; }
  inline long getNumResizes() const { return num_resizes; }
protected:
  inline long nextSize() const { return SIZES[size_index+1]; }
  inline long prevSize() const { return size_index ? SIZES[size_index-1] : 0; }
  inline void Enlarge() {
    void** newt = (void**) realloc(table, nextSize() * sizeof(void*));
    if (0==newt) return;
    table = newt;
    for (long i=getSize(); i<nextSize(); i++) table[i] = 0;
    size_index++;
    if (size_index > max_size_index) max_size_index = size_index;
    num_resizes++;
  };
  inline void Shrink() {
    void** newt = (void**) realloc(table, prevSize() * sizeof(void*));
    if (0==newt) return;
    table = newt;
    size_index--;
    num_resizes++;
  }
};

const long voidhash::SIZES[] = {
    11, 23, 47, 97, 197, 397, 797, 1597, 3203, 6421, 12853, 25717, 51437,
    102877, 205759, 411527, 823117, 1646237, 3292489, 6584983, 13169977,
    26339969, 52679969, 105359939, 210719881, 421439783, 842879579, 1685759167,
    -1
};

/** Hash table template class.

    To use this template, you must define the following member functions
    for the class to store objects in the hash table.

    \begin{description}
 
    \item[long Signature(long M) const]  
    The hash function.  M is the size of the hash table.  Should return 
    something between 0 and M-1, where M is guaranteed to be prime.

    \item[bool Equals(DATA\_TYPE *) const]  
    So we can catch equal objects in the hash table.  Note that equal objects
    (according to Equals) MUST have the same signature regardless of M.  

    \item[void SetNext(DATA\_TYPE *)]
    The hash table uses chaining, and the objects must store the next pointer.
    This method sets the next pointer.

    \item[DATA\_TYPE* GetNext() const]
    Get the next pointer.

    \end{description}

*/
template <class DATA_TYPE>
class HashTable : public voidhash {
public:
  HashTable() : voidhash() { }

  /// Find an object equal to the one specified; if none present, return 0.
  DATA_TYPE* Find(const DATA_TYPE* key);

  /** Insert an object into the table.
      If a duplicate object is already present, the original is not inserted
      and the duplicate is returned.  Otherwise, the original is inserted
      and is returned.
  */
  DATA_TYPE* UniqueInsert(DATA_TYPE*);

  /// Removes the object specified, if present.
  void Remove(DATA_TYPE*);
  
protected:
  inline bool MoveToFront(long slot, const DATA_TYPE* key) {
    DATA_TYPE* list = (DATA_TYPE*) table[slot];
    if (0==list) return false;
    long c = 0;
    DATA_TYPE* prev = 0;
    DATA_TYPE* curr = list;
    for (; curr; curr=curr->GetNext()) {
      c++;
      if (key->Equals(curr)) break;
      prev = curr;
    }
    maxchain = MAX(maxchain, c);
    if (0==curr) return false;
    if (prev) {
      prev->SetNext(curr->GetNext());
      curr->SetNext(list);
      table[slot] = curr;
    }
    return true;
  };

  DATA_TYPE* ConvertToList();
  void BuildFromList(DATA_TYPE*);
};


template <class DT>
DT* HashTable <DT> :: Find(const DT* key)
{
  long slot = key->Signature(getSize());
  if (MoveToFront(slot, key)) {
    return (DT*) table[slot];
  } 
  return 0;
}

template <class DT>
DT* HashTable <DT> :: UniqueInsert(DT* key)
{
  long slot = key->Signature(getSize());
  if (MoveToFront(slot, key)) {
    return (DT*) table[slot];
  }
  key->SetNext((DT*)table[slot]);
  table[slot] = key;
  elements++;
  maxelems = MAX(maxelems, elements);
  if (nextSize()>0 && elements > nextSize()) {
    DT* list = ConvertToList();
    Enlarge();
    BuildFromList(list);
  }
  return key;
}

template <class DT>
void HashTable <DT> :: Remove(DT* key)
{
  long slot = key->Signature(getSize());
  if (!MoveToFront(slot, key)) return;
  if (table[slot] != key) return;
  table[slot] = key->GetNext();
  key->SetNext(0);
  elements--;
  if (elements*2 < prevSize()) {
    DT* list = ConvertToList();
    Shrink();
    BuildFromList(list);
  }
}

template <class DT>
DT* HashTable <DT> :: ConvertToList()
{
  DT* biglist = 0;
  long i = getSize();
  for (i--; i>=0; i--) {
    DT* list = (DT*) table[i];
    table[i] = 0;
    while (list) {
      DT* next = list->GetNext();
      list->SetNext(biglist);
      biglist = list;
      list = next;
    }
  }
  return biglist;
}

template <class DT>
void HashTable <DT> :: BuildFromList(DT* list)
{
  while (list) {
    long slot = list->Signature(getSize());
    DT* next = list->GetNext();
    list->SetNext((DT*)table[slot]);
    table[slot] = list;
    list = next;
  }
}

#endif
