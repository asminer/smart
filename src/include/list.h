
#ifndef LIST_H
#define LIST_H

#include "defines.h"
#include <stdlib.h>  // for realloc
#include <string.h>  // for memcpy

class voidlist {
  void** data;
  int alloc;
  int last;
protected:
  inline bool Resize(int ns) {
    if (ns == alloc)  return 1;
    void** newdata = (void**) realloc(data, ns * sizeof(void*));
    if (ns) {
      if (0==newdata)  return 0;
      data = newdata;
      alloc = ns;
    } 
    return 1;
  }
public:
  voidlist() {
    data = 0;
    alloc = 0;
    last = 0;
  }
  ~voidlist() {
    free(data);
  }
  inline int Length() const { return last; }
  inline bool Append(void* x) {
    if (last >= alloc) {
      int ns;
      if (alloc < 1)          ns = 16;
      else if (alloc < 1024)  ns = alloc*2;
      else                    ns = alloc + 1024;
      if (!Resize(ns))        return false;
    }
    data[last] = x;
    last++;
    return true;
  }
  inline void* Item(int n) {
    CHECK_RANGE(0, n, last);
    return data[n];
  }
  inline const void* ReadItem(int n) const {
    CHECK_RANGE(0, n, last);
    return data[n];
  }
  inline void Clear() {
    last = 0;
  }
  inline void Compact() {
    Resize(last);
  }
  inline void** CopyIntoArray() const {
    if (0==last)  return 0;
    void** foo = (void**) malloc(last * sizeof(void*));
    memcpy(foo, data, last * sizeof(void*));
    return foo;
  }
  inline void** CopyAndClear() {
    Compact();
    void** foo = data;
    last = 0;
    data = 0;
    alloc = 0;
    return foo;
  }
};

template <class DATA>
class List {
  voidlist L;
public:
  inline int Length() const { 
    return L.Length(); 
  }
  inline bool Append(DATA* x) { 
    return L.Append(x); 
  }
  inline DATA* Item(int n) { 
    return (DATA*) L.Item(n); 
  }
  inline const DATA* ReadItem(int n) const {
    return (const DATA*) L.ReadItem(n);
  }
  inline void Clear() {
    L.Clear();
  }
  inline void Compact() {
    L.Compact();
  }
  inline DATA** CopyIntoArray() const {
    return (DATA**) L.CopyIntoArray();
  }
  inline DATA** CopyAndClear() {
    return (DATA**) L.CopyAndClear();
  }
};


template <class DATA>
class ObjectList {
  DATA* data;
  int alloc;
  int last;
protected:
  inline bool Resize(int ns) {
    if (ns == alloc)  return 1;
    DATA* newdata = (DATA*) realloc(data, ns * sizeof(DATA));
    if (ns) {
      if (0==newdata)  return 0;
      data = newdata;
      alloc = ns;
    } 
    return 1;
  }
public:
  ObjectList() {
    data = 0;
    alloc = 0;
    last = 0;
  }
  ~ObjectList() {
    free(data);
  }
  inline int Length() const { return last; }
  inline bool Append(DATA x) {
    if (last >= alloc) {
      int ns;
      if (alloc < 1)    ns = 16;
      else if (alloc < 1024)  ns = alloc*2;
      else      ns = alloc + 1024;
      if (!Resize(ns))    return false;
    }
    data[last] = x;
    last++;
    return true;
  }
  inline DATA Item(int n) const {
    CHECK_RANGE(0, n, last);
    return data[n];
  }
  inline void Clear() {
    last = 0;
  }
  inline void Compact() {
    Resize(last);
  }
  inline DATA* CopyIntoArray() const {
    if (0==last)  return 0;
    DATA* foo = (DATA*) malloc(last * sizeof(DATA));
    memcpy(foo, data, last * sizeof(DATA));
    return foo;
  }
  inline DATA* CopyAndClear() {
    Compact();
    DATA* foo = data;
    last = 0;
    data = 0;
    alloc = 0;
    return foo;
  }
};


#endif
