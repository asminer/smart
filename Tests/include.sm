
// $Id$

// Test of includes and line numbers

print("Starting in file ", file_name, "; currently on line ", line_number, " (6)\n");
#include "inc-a.sm"
print("Back to file ", file_name, "; on line ", line_number, " (8)\n");
#include "inc-b.sm"
print("Back to file ", file_name, "; on line ", line_number, " (10)\n");

print("Circular include, should cause an error:\n");
#include "include.sm"

print("done; last line of file is ", line_number, "\n");
