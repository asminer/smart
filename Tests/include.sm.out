Starting in file include.sm; currently on line 6 (6)
2-lines; file is inc-a.sm; on line 2 (2)
Back to file include.sm; on line 8 (8)
in file inc-b.sm; on line 2 (2)
2-lines; file is inc-a.sm; on line 2 (2)
leaving file inc-b.sm; on line 4 (4)
Back to file include.sm; on line 10 (10)
Circular include, should cause an error:
WARNING in file include.sm near line 13:
	circular file dependency caused by #include include.sm, ignoring
done; last line of file is 15
